plugins/arElasticSearchPlugin/lib/model/arElasticSearchInformationObjectPdo.class.php

```diff
@@ -612,32 +619,80 @@ public function getNameAccessPoints()
   }
 
   /*
-   * Get related terms and its ancestors
+   * Get related terms plus any ancestors
    */
   protected function getRelatedTerms($typeId)
   {
+    $relatedTerms = array();
+
     $sql  = 'SELECT
-                DISTINCT term.id,
-                term.taxonomy_id,
-                term.source_culture,
-                slug.slug,
-                i18n.name';
+                DISTINCT current.id';
     $sql .= ' FROM '.QubitObjectTermRelation::TABLE_NAME.' otr';
     $sql .= ' JOIN '.QubitTerm::TABLE_NAME.' current
                 ON otr.term_id = current.id';
-    $sql .= ' JOIN '.QubitTerm::TABLE_NAME.' term
-                ON term.lft <= current.lft AND term.rgt >= current.rgt';
-    $sql .= ' JOIN '.QubitTermI18n::TABLE_NAME.' i18n
-                ON term.id = i18n.id';
-    $sql .= ' JOIN '.QubitSlug::TABLE_NAME.' slug
-                ON term.id = slug.object_id';
     $sql .= ' WHERE otr.object_id = ?
-               AND term.taxonomy_id = ?';
+                AND current.taxonomy_id = ?';
 
     self::$statements['relatedTerms'] = self::$conn->prepare($sql);
     self::$statements['relatedTerms']->execute(array($this->__get('id'), $typeId));
 
-    return self::$statements['relatedTerms']->fetchAll(PDO::FETCH_OBJ);
+    // Get directly related terms.
+    $rows = self::$statements['relatedTerms']->fetchAll(PDO::FETCH_ASSOC);
+
+    if (0 == count($rows))
+    {
+      return $relatedTerms;
+    }
+
+    // Iterate over each directly related term, adding all ancestors of each
+    foreach($rows as $row)
+    {
+      $relatedTerms = array_merge($relatedTerms, $this->recursivelyGetParentTerms(array($row['id'])));
+    }
+
+    // Merge unique related terms to $rows
+    $relatedTerms = array_unique($relatedTerms);
+    sort($relatedTerms, SORT_NUMERIC);
+
+    return $relatedTerms;
+  }
+
+  /**
+   * Recursively find all parent terms for any given term. Subject, Place and
+   * Genre terms are preloaded in array $this->terms. The import array $ids is
+   * appended to for every parent that is added on each recursive call.
+   *
+   * @param array $ids  An array with one term id populated.
+   * @return array  The input array with directly related term first, followed by
+   *                any parents found, in child to top parent order.
+   */
+  private function recursivelyGetParentTerms($ids)
+  {
+    // If there were no terms loaded, then there are no ancestors.
+    if (0 >= count($this->terms))
+    {
+      return $ids;
+    }
+
+    // Error in input. Return empty array.
+    if (!(is_array($ids) && 0 < count($ids)))
+    {
+      return array();
+    }
+
+    // Get last term id found in array.
+    $last_id = end($ids);
+    // Get immediate parent of last_id from $this->terms.
+    $parent = $this->terms[$last_id];
+
+    // Do not include root term id.
+    if (QubitTerm::ROOT_ID != $parent)
+    {
+      $ids[] = $parent;
```

Comment:
I think this recursive function could be much simpler with a pattern like:

```
if (QubitTerm::ROOT_ID == $parent)
{
  return array($id);
} 
else 
{
  return array_unshift($this->recursivelyGetParentTerms($parent), $id);
}
```

This would avoid the need to pass an array of $ids to recursivelyGetParentTerms().

Reply:
Good idea - this is a better pattern!  I have refactored the code.